textile/chat-features.textile

```diff
@@ -9,5 +9,148 @@ jump_to:
+** @(CHA-V1a)@ The current version of the Chat library feature specification is version @0.1.0@.
+
+h2(#general). General Principles
+
+* @(CHA-GP1)@ As far as is practicable, the implementation details such as underlying Realtime Channels should be hidden from the public API. This allows developers to interact with Chat without having to understand many low-level primitives.
+* @(CHA-GP2)@ The public API should avoid implicit operations as a side-effect to some other operation. For example, adding a subscriber to messages in a Chat Room should not automatically trigger that Room to attach. This is in contrast to the current core SDKs. Avoiding side-effects provides a clean, easy to understand API.
+* @(CHA-GP3)@ Wherever possible, Chat features should be exposed in the public API as properties of their parent. For example, @messages@ would be considered a property of a @room@ object. This allows for greater composability and extensibility in the future.
+
+h2(#rooms). Rooms
+
+h3(#rooms-general). General Information
+
+A Room is the atomic unit of Chat. All chat operations are performed in the context of a room that a client is currently attached to. As the Room is the atomic unit, its state should be considered as the combination of the states of underlying resources.
+
+h3(#rooms-status). Room Statuses
+
+The status of any given Chat Room is the combination of the states of all of its constituent contributors. For more information on this, see "@Room Lifecycle@":#rooms-lifecycle.
+
+* @(CHA-RS1)@ Every room has a @status@, which describes the current status of the room. 
+** @(CHA-RS1a)@ The @INITIALIZED@ status is the initial status before any attachment operations have taken place.
+** @(CHA-RS1b)@ The @ATTACHING@ status is used when the room is in the process of attaching to Ably.
+** @(CHA-RS1c)@ The @ATTACHED@ status means that the room is fully connected to Ably and functioning normally.
+** @(CHA-RS1d)@ The @DETACHING@ status is used when the room is in the process of detaching from Ably.
+** @(CHA-RS1e)@ The @DETACHED@ status means that the room has been detached from Ably by a user-initiated request. It will not attempt to re-connect without explicit intervention.
+** @(CHA-RS1f)@ The @SUSPENDED@ status is when the room has been detached from Ably for an extended period of time. The room will periodically attempt to reconnect.
+** @(CHA-RS1g)@ The @FAILED@ status means that something has gone wrong with the room (such as lack of permissions). The room will not attempt to connect to Ably and will require user intervention.
+** @(CHA-RS1h)@ The @RELEASING@ status means that the room is being released and the underlying resources are being cleaned up.
+** @(CHA-RS1i)@ The @RELEASED@ status means that the room has been cleaned up and the object can no longer be used.
+* @(CHA-RS2)@ A room must expose its current status.
+** @(CHA-RS2a)@ @[Testable]@ A room must expose its current status, a single value from the list provided above.
```

Comment:
Was it a conscious decision to not describe the API of the SDK in terms of signatures / types (e.g. the way that we do in the main spec, which also has the IDL)?

Reply:
The IDL we have for the core SDKs is quite java-centric and in that sense may not be the most idiomatic approach for other ecosystems, so I tried to write the spec in a way that describes the intended interactions and behaviour, rather than prescribe an exact implementation.

Is it worth a note somewhere in the spec to treat the JS implementation as a reference implementation and explain the intention?

Reply:
I believe having IDL for chat-SDK will be super helpful. 
Almost all of the languages these days are encouraging use of types. 
So, having language-agnostic types will surely help.